cpp/src/parquet/arrow/reader.cc

```diff
@@ -1107,8 +1112,13 @@ class RowGroupGenerator {
     if (!reader->properties().pre_buffer()) {
       row_group_read = SubmitRead(cpu_executor_, reader, row_group, column_indices);
     } else {
+      reader->parquet_reader()->PreBuffer({row_group}, column_indices_,
+                                          reader_properties_.io_context(),
+                                          reader_properties_.cache_options());
       auto ready = reader->parquet_reader()->WhenBuffered({row_group}, column_indices);
       if (cpu_executor_) ready = cpu_executor_->TransferAlways(ready);
+      // wait for buffering the rowgroup before read it
+      ready.Wait();
```

Comment:
Hmm, why are we synchronously blocking, then attaching a callback to the future in the next step? Something seems off here

Reply:
Thanks David, I got this error when I don't use wait:
`ReadRangeCache did not find matching cache entry`
It's like the `parquet_reader()->WhenBuffered` call is not really waiting to have the buffer ready before read the rowgroup.
I'll investigate more what could be the issue here.

Reply:
It seems -somehow- the next rowgroup triggers `ReadOneRowGroup` and it starts reading, but the cache entries are empty (because it didn't call `PreBuffer` before)
I did a simple change: when we don't transfer the future using
`if (cpu_executor_) ready = cpu_executor_->TransferAlways(ready);`
it works without wait.

Reply:
I think that when we transfer the future, the `PreBuffer` that it's running now (in a different thread) cannot finish before the future triggers `ReadOneRowGroup` (because the future was already transferred into a different thread)
So perhaps we just  need a way to make sure `PreBuffer` gets called before in the same thread of the future.

Reply:
I sent [new changes](https://github.com/apache/arrow/pull/14226/files#diff-6a73dc8ee650cf7b5b8a51f7d4f9f0197823416155c0de7de400294fea588f11R1118) to avoid the use of `Future::Wait` and to run the prebuffering and read operations together in the same thread context (one after the other).